More safety changes for "Effective Dart". (#2986)
dart-lang/site-www: src/_guides/language/effective-dart/design_migrated.md

your default should be to make fields and top-level variables `final` when you
can.

Sometimes an instance field doesn't change after it has been initialized, but
can't be initialized until after the instance is constructed. For example, it
may need to reference `this` or some other field on the instance. In cases like,

> that, consider making the field `late final`.

Reviewer: ```suggestion
that, consider making the field `late final`. When a field is `late final` it may also be possible to use an initializing expression on the field declaration.
```

Author: Good suggestion. I'll reword this a little and apply it to my change.

[angular]: {{site.angulardart}}

### AVOID public `late final` fields.

> Unlike other final fields, a late final field *does* define a setter. If

Reviewer: ```suggestion
Unlike other final fields, a late final field without an initializing expression *does* define a setter. If
```

Author: 😮 Today I learned.

Reworded and updated the guideline title too.

that field is public, then the setter is public. This is rarely what you want.
Fields are usually marked late so that they can be initialized *internally* at
some point in the instance's lifetime, often inside the constructor body.

Unless you do intend to expose the setter publicly and allow users to invoke it,
it's better to either not use `late`, or make the late field private with a
public getter.

### AVOID returning nullable `Future`, `Stream`, and collection types.

When an API returns a container type, it has two ways to indicate the absence of
data: It can return an empty container or it can return `null`. Users generally
assume and prefer that you use an empty container to indicate "no data". That
way, they have a real object that they can call methods on like `isEmpty`.

> Prefer returning an empty collection, a future that doesn't complete, or a

Reviewer: A future that doesn't complete is not often what users would want. More likely you want a non-nullable future with a nullable generic.

Author: Ah, good call. Done.

[angular]: {{site.angulardart}}

### AVOID public `late final` fields without initializers.

Unlike other `final` fields, a `late final` field without an initializer *does*
define a setter. If that field is public, then the setter is public. This is
rarely what you want. Fields are usually marked late so that they can be
initialized *internally* at some point in the instance's lifetime, often inside
the constructor body.

> Unless you do want users to call the setter, it's better to either:
> 
> * not use `late`,
> * ensure the `late` field is initialized at its declaration,
> * or make the `late` field private and define a public getter to access it.

Reviewer 2: So much easier to read! Some further tweaks... Apparently, "either <3 things>" is controversial (https://developers.google.com/style/word-list#letter-e). Also, list items should be parallel. So maybe:

```suggestion
Unless you *do* want users to call the setter, it's better to pick one of the
following solutions:

* Don't use `late`.
* Use `late`, but initialize the `late` field at its declaration.
* Use `late`, but make the `late` field private and define a public getter for it.
```
